src/main/java/emu/grasscutter/game/avatar/Avatar.java

```diff
@@ -779,6 +779,15 @@ public boolean setSkillLevel(int skillId, int level) {
         return true;
     }
 
+    public void changeSkillLevel(int skillLevel) {
+        if (skillLevel < 1) skillLevel = 1;
+        if (skillLevel > 12) skillLevel = 12; // Set max to 12 at the moment instead of 15 (it will conflict with constellation's talent level bonus)
```

Comment:
Check `Avatar::setSkillLevel`, constellation bonuses have also been capped in 4e75361 which also correctly caps other talents (e.g. alt sprints) so it would be safe (even if slightly incorrect) to just use `AvatarSkillDepotData::getSkillsAndEnergySkill` elsewhere.

Reply:
>also using `Avatar::setSkillLevel` leads to `"emu.grasscutter.game.avatar.Avatar.getPlayer()" is null`.

Ah right, won't have a player during construction. Should be fine to add a conditional block to the packets in it.
```java
        // Packet
        val player = this.getPlayer();
        if (player != null) {
            player.sendPacket(new PacketAvatarSkillChangeNotify(this, skillId, oldLevel, level));
            player.sendPacket(new PacketAvatarSkillUpgradeRsp(this, skillId, oldLevel, level));
        }
```

> Also for the `AvatarSkillDepotData::getSkillsAndEnergySkill` does it mean I should revert my changes for `AvatarSkillDepotData` and `TalentCommand`?

Up to you, might be useful regardless and it's a pretty harmless addition.

> I added `Avatar::changeSkillLevel` specifically for my implementation since I only needed skillLevel as input 

This is a bit overspecialised though.